toothpick-runtime/src/main/java/toothpick/ScopeImpl.java

```diff
@@ -272,69 +270,60 @@ private void installModule(Module module) {
    * in the current scope.
    *
    * @param clazz the class for which to install the scoped provider of this scope.
    * @param bindingName the name, possibly {@code null}, for which to install the scoped provider.
+   * @param provider the provider that will be used to obtain the injected instances.
    * @param <T> the type of {@code clazz}.
    */
-  private <T> void installScopedProvider(Class<T> clazz, String bindingName, ScopedProviderImpl<? extends T> provider) {
+  private <T> void installScopedProvider(Class<T> clazz, String bindingName, InternalProviderImpl<? extends T> provider) {
     synchronized (clazz) {
       UnNamedAndNamedProviders<T> unNamedAndNamedProviders = mapClassesToAllProviders.get(clazz);
       if (unNamedAndNamedProviders == null) {
         unNamedAndNamedProviders = new UnNamedAndNamedProviders<>();
         mapClassesToAllProviders.put(clazz, unNamedAndNamedProviders);
       }
+
       if (bindingName == null) {
-        unNamedAndNamedProviders.setUnNamedProvider(provider);
+        unNamedAndNamedProviders.unNamedProvider = provider;
       } else {
-        Map<String, ScopedProviderImpl<? extends T>> mapNameToProvider = unNamedAndNamedProviders.getMapNameToProvider();
-        if (mapNameToProvider == null) {
-          mapNameToProvider = new HashMap<>();
-          unNamedAndNamedProviders.setMapNameToProvider(mapNameToProvider);
-        }
-        mapNameToProvider.put(bindingName, provider);
+        unNamedAndNamedProviders.putNamedProvider(bindingName, provider);
       }
     }
   }
 
-  /**
-   * Install the unScopedProvider of the class {@code clazz}
-   * in the pool of unscoped providers.
-   *
-   * @param clazz the class for which to install the unscoped unScopedProvider.
-   * @param <T> the type of {@code clazz}.
-   */
-  private <T> void installUnScopedProvider(Class<T> clazz, UnScopedProviderImpl<? extends T> unScopedProvider) {
-    synchronized (clazz) {
-      mapClassesToUnScopedProviders.put(clazz, unScopedProvider);
-    }
-  }
-
   private static class UnNamedAndNamedProviders<T> {
-    private ScopedProviderImpl<? extends T> unNamedProvider;
-    private Map<String, ScopedProviderImpl<? extends T>> mapNameToProvider;
-
-    public Map<String, ScopedProviderImpl<? extends T>> getMapNameToProvider() {
-      return mapNameToProvider;
-    }
+    private InternalProviderImpl<? extends T> unNamedProvider;
+    private Map<String, InternalProviderImpl<? extends T>> mapNameToProvider;
 
-    public void setMapNameToProvider(Map<String, ScopedProviderImpl<? extends T>> mapNameToProvider) {
-      this.mapNameToProvider = mapNameToProvider;
-    }
-
-    public void setUnNamedProvider(ScopedProviderImpl<? extends T> unNamedProvider) {
-      this.unNamedProvider = unNamedProvider;
+    private void putNamedProvider(String name, InternalProviderImpl<? extends T> provider) {
+      if (mapNameToProvider == null) {
+        mapNameToProvider = new HashMap<>();
+      }
```

Comment:
ok this will need quite some work after the lock free PR. To make it easier, do not lazy the creation of the map.